src/frontend/src/optimizer/plan_node/stream_sink.rs

```diff
@@ -465,46 +466,63 @@ impl StreamSink {
         format_desc: Option<&SinkFormatDesc>,
     ) -> Result<SinkType> {
         let frontend_derived_append_only = input_append_only;
-        let (user_defined_append_only, user_force_append_only, syntax_legacy) = match format_desc {
+        let (user_defined_sink_type, user_force_append_only, syntax_legacy) = match format_desc {
             Some(f) => (
-                f.format == SinkFormat::AppendOnly,
+                Some(match f.format {
+                    SinkFormat::AppendOnly => SinkType::AppendOnly,
+                    SinkFormat::Upsert | SinkFormat::Debezium => SinkType::Upsert,
+                }),
                 Self::is_user_force_append_only(&WithOptionsSecResolved::without_secrets(
                     f.options.clone(),
                 ))?,
                 false,
             ),
             None => (
-                Self::is_user_defined_append_only(properties)?,
+                Self::sink_type_in_prop(properties)?,
                 Self::is_user_force_append_only(properties)?,
                 true,
             ),
         };
 
-        match (
-            frontend_derived_append_only,
-            user_defined_append_only,
-            user_force_append_only,
-        ) {
-            (true, true, _) => Ok(SinkType::AppendOnly),
-            (false, true, true) => Ok(SinkType::ForceAppendOnly),
-            (_, false, false) => Ok(SinkType::Upsert),
-            (false, true, false) => {
-                Err(ErrorCode::SinkError(Box::new(Error::new(
+        if let Some(user_defined_sink_type) = user_defined_sink_type {
+            if user_defined_sink_type == SinkType::AppendOnly {
+                if user_force_append_only {
+                    return Ok(SinkType::ForceAppendOnly);
```

Comment:
When upstream is append-only, and user specifies both append-only and force-append-only, the `SinkType` changed from `AppendOnly` to `ForceAppendOnly`.